Adds QOI support to Peek

pedrolamas opened the pull request:
<!-- Enter a brief description/summary of your PR here. What does it fix/what does it change/how was it tested (even manually, if necessary)? -->
## Summary of the Pull Request

Adds QOI image files support to Peek!

### Light theme

![image](https://github.com/microsoft/PowerToys/assets/85504/02a0701c-2b9c-450b-844a-8173b1c20626)

### Dark theme

![image](https://github.com/microsoft/PowerToys/assets/85504/0d54f292-fa12-4026-9747-18adeec0051c)

<!-- Please review the items on the PR checklist before submitting-->
## PR Checklist

- [X] **Closes:** #29918
- [X] **Communication:** I've discussed this with core contributors already. If work hasn't been agreed, this work might be rejected
- [X] **Tests:** Added/updated and all pass
- [ ] **Localization:** All end user facing strings can be localized
- [ ] **Dev docs:** Added/updated
- [ ] **New binaries:** Added on the required places
   - [ ] [JSON for signing](https://github.com/microsoft/PowerToys/blob/main/.pipelines/ESRPSigning_core.json) for new binaries
   - [ ] [WXS for installer](https://github.com/microsoft/PowerToys/blob/main/installer/PowerToysSetup/Product.wxs) for new binaries and localization folder
   - [ ] [YML for CI pipeline](https://github.com/microsoft/PowerToys/blob/main/.pipelines/ci/templates/build-powertoys-steps.yml) for new test projects
   - [ ] [YML for signed pipeline](https://github.com/microsoft/PowerToys/blob/main/.pipelines/release.yml)
- [ ] **Documentation updated:** If checked, please file a pull request on [our docs repo](https://github.com/MicrosoftDocs/windows-uwp/tree/docs/hub/powertoys) and link it here: #xxx

<!-- Provide a more detailed description of the PR, other things fixed or any additional comments/features here -->
## Detailed Description of the Pull Request / Additional comments

A couple of notes on this PR:

- I moved the `QoiImage` and `Gcode*` classes from PreviewHandlerCommon to FilePreviewCommon, and by doing that, I had to enable `<UseWindowsForms>`, `<AllowUnsafeBlocks>` and `<Nullable>` on the project.
- As I enabled nullable types on the step above, I fixed all the nullable issues on FilePreviewCommon
- FilePreviewCommon has been added as a project dependency to:
  - GcodePreviewHandler
  - GcodeThumbnailProvider
  - QoiPreviewHandler
  - QoiThumbnailProvider
- `IFileSystemItemExtensions.GetQoiSize` is just going directly to the bytes where we expect the "width" and "height" of a QOI image to be. I did not add a file header validation, but that should be rather trivial to do as it is already implemented in `QoiImage`.
- `BitmapHelper.GetBitmapFromHBitmapAsync` has been refactored so that I could extract the code that converts a `System.Drawing.Bitmap` to a `Microsoft.Xaml...BitmapSource` **<- you might want to take a good look at this one to make sure all is ok!**

I understand that Peek can now show Preview Pane Handlers directly, but considering #26726, I think it best to prepare to a situation where PT stops supporting the Preview Pane Handlers completely

<!-- Describe how you validated the behavior. Add automated tests wherever possible, but list manual validation steps taken as well -->
## Validation Steps Performed

Manual validation with the test images pack in https://qoiformat.org/

htcfreek commented:
@pedrolamas 
Isn't it ready? Then we can remove the review label. Thought it was ready.

pedrolamas commented:
@htcfreek from my part, this PR is ready to go!

jaimecbernardo commented:
Thanks for opening the PR, @pedrolamas .
At this moment, we're thinking of reviewing the contribution for 0.77 only. Sorry for the delay. 😄

pedrolamas commented:
No problem, thanks for letting me know, @jaimecbernardo 

I will try and keep the branch in sync with main branch to make sure it does not have conflicts.

stefansjfw commented:
/azp run

jaimecbernardo reviewed (changes requested):
Hi @pedrolamas ,
Thank you for the contribution!
I've tested and it seems to look good, but it seems to bring a regression:
![image](https://github.com/microsoft/PowerToys/assets/26118718/f4690198-32cb-40c9-a24d-c2b860349449)
When the preview pane is on a qoi file, you can't delete that file. This was OK in 0.76.0, it seems. Can you please look into it? I suspect the refactor caused some resources to not be freed correctly.

jaimecbernardo commented:
I've merged latest main in and added QOI to the notice section about Peek as well. Running a Dart CI test to check if it still builds OK for release.

pedrolamas commented:
> When the preview pane is on a qoi file, you can't delete that file.

Thanks @jaimecbernardo, I have a feeling this is due to the fact I changed the `BinaryReader` in `QoiReader` to leave the stream open (in theory that makes sense to me as I would expect the stream to be closed upstream, but maybe that is not the case?)

pedrolamas commented:
@jaimecbernardo I have now fixed the problem by adding a `using` to the file stream

https://github.com/microsoft/PowerToys/pull/29919/files#diff-1d420d0e6cfb94016296228da268defbf9677c8414f378b21b45abb19ce22779R66

For the record, my thinking here is that if the `Image.FromStream` method leaves the stream open, I think it is reasonable to expect the same behavior from `QoiImage.FromStream`, hence why I changed the `BinaryReader` to leave it open.

Disposing the `FileStream` in this case solves the problem very nicely, and I would argue that it is reasonable that (where appropriate) the same change should be made on ALL other preview handlers, what do you think?

jaimecbernardo commented:
Solves it for me. Not sure 🤔 . The other previewers don't seem to be suffering from the same issue.

jaimecbernardo reviewed (approved):
LGTM! Thank you for the contribution!
